processor/transformer/transformer.go

```diff
@@ -86,35 +94,48 @@ type HandleT struct {
 	receivedStat stats.Measurement
 	cpDownGauge  stats.Measurement
 
+	conf   *config.Config
 	logger logger.Logger
+	stat   stats.Stats
 
 	Client *http.Client
 
 	guardConcurrency chan struct{}
+
+	config struct {
+		maxConcurrency         int
+		maxHTTPConnections     int
+		maxHTTPIdleConnections int
+		disableKeepAlives      bool
+
+		timeoutDuration time.Duration
+
+		maxRetry int
+
+		failOnTimeout bool
+		failOnError   bool
+
+		destTransformationURL string
+		userTransformationURL string
+	}
+
+	warehouseDestinationMap map[string]struct{}
 }
 
 // Transformer provides methods to transform events
 type Transformer interface {
-	Setup()
-	Transform(ctx context.Context, clientEvents []TransformerEventT, url string, batchSize int) ResponseT
-	Validate(clientEvents []TransformerEventT, url string, batchSize int) ResponseT
+	Setup(*config.Config, logger.Logger, stats.Stats)
+	Transform(ctx context.Context, clientEvents []TransformerEventT, batchSize int) ResponseT
+	UserTransform(ctx context.Context, clientEvents []TransformerEventT, batchSize int) ResponseT
+	Validate(ctx context.Context, clientEvents []TransformerEventT, batchSize int) ResponseT
 }
 
 // NewTransformer creates a new transformer
 func NewTransformer() *HandleT {
 	return &HandleT{}
 }
```

Comment:
What is the value of producing an uninitialised transformer that is unusable unless you call Setup first?

I would much prefer to have all transformer.Handle fields unexported and a single producer method:

```go
type Opt func(*handle)

func NewTransformer(conf *config.Config, log logger.Logger, stat stats.Stats, opts... Opts) Transformer
```

I understand that this requires further refactoring though